src/dead_code_elimination.cpp

```diff
@@ -60,7 +60,10 @@ void dead_code_elimination::apply(module& m) const
                 leaf->clear_arguments();
                 assert(bidistance(m, last, leaf) < 0);
                 assert(leaf != ins);
-                m.move_instruction(leaf, m.end());
+                if(leaf->name() != "@param" or
+                   (leaf->name() == "@param" and
+                    any_cast<builtin::param>(leaf->get_operator()).parameter == "scratch"))
```

Comment:
Why are we checking the name?

Reply:
After some investigation, the problem is, in the "proallocate_param" pass, the scratch memory is replaced by the preallocate_param instruction as:

```
        std::string id = m.name() + ":" + param;
        auto r         = m.insert_instruction(ins, model.preallocate(ins->get_shape(), id));
        m.replace_instruction(ins, r);
```
So, we do not need the scratch parameter anymore. But with the change in the dead code elimination, we cannot remove the scratch parameter, then when we run a program on the GPU, there is a scratch parameter that is not provided, so an error is reported.

I can change this pass to remove the scratch param right after the replace_instruction as:

```
void preallocate_param::apply(module& m) const
{
    auto last = std::prev(m.end());
    for(auto ins : iterator_for(m))
    {
        if(ins->name() != "@param")
            continue;
        if(param != any_cast<builtin::param>(ins->get_operator()).parameter)
            continue;
        std::string id = m.name() + ":" + param;
        auto r         = m.insert_instruction(ins, model.preallocate(ins->get_shape(), id));
        m.replace_instruction(ins, r);
        m.move_instruction(ins, m.end());
    }
    m.remove_instructions(std::next(last), m.end());
}
```

Then we do not need to check the parameter name in the dead code elimination anymore.

Reply:
> I can change this pass to remove the scratch param right after the replace_instruction

I think thats a better change.